src/compute-client/src/controller/instance.rs

```diff
@@ -797,31 +808,36 @@ where
                 compute_net.push((key, update));
             } else {
                 // Storage presumably, but verify.
-                storage_todo
-                    .entry(key)
-                    .or_insert_with(ChangeBatch::new)
-                    .extend(update.drain())
+                if self.storage_controller.collection(key).is_ok() {
+                    storage_todo
+                        .entry(key)
+                        .or_insert_with(ChangeBatch::new)
+                        .extend(update.drain())
+                }
             }
         }
 
-        // Translate our net compute actions into `AllowCompaction` commands.
+        // Translate our net compute actions into `AllowCompaction` commands
+        // and a list of collections that are potentially ready to be dropped
         let mut compaction_commands = Vec::new();
+        let mut dropped_collection_ids = Vec::new();
         for (id, change) in compute_net.iter_mut() {
+            let frontier = self.compute.collection(*id)?.read_frontier();
+            if frontier.is_empty() {
+                dropped_collection_ids.push(id.clone());
+            }
```

Comment:
I'm not sure if this part works as you expect. AFAICT, `compute_net` will only include changes that advance the global read frontier of the collection. So if that doesn't advance (e.g. the slower of two replica reports the empty frontier for a collection), this loop won't be entered so the drop check is not performed for the collection.

Reply:
I am not sure I understand this comment. The general approach is to try to cover changes to the read frontier of the collection as well as changes to its `replica_write_frontiers` map. That way, whenever the garbage collection condition has a chance to turn true, we check it.

Reply:
I think I confused myself here a bit, thinking we are inside `update_write_frontiers`, but this is actually `update_read_capabilities`. My point is still valid, I think, namely that we don't cover all changes to `replica_write_frontiers` here, only the ones that also result in changes to the collection read frontier. Which might be fine, considering that we also have a check in `handle_frontier_uppers`, which does see all replica frontier updates.